src/core/datagram.c

```diff
@@ -0,0 +1,434 @@
+        QUIC_TEL_ASSERT(Datagram->Enabled);
+
+        if (SendRequest->Flags & QUIC_SEND_FLAG_DGRAM_PRIORITY) {
+            SendRequest->Next = *Datagram->PrioritySendQueueTail;
+            *Datagram->PrioritySendQueueTail = SendRequest;
+            if (Datagram->SendQueueTail == Datagram->PrioritySendQueueTail) {
+                Datagram->SendQueueTail = &SendRequest->Next;
+            }
+            Datagram->PrioritySendQueueTail = &SendRequest->Next;
+        } else {
+            *Datagram->SendQueueTail = SendRequest;
+            Datagram->SendQueueTail = &SendRequest->Next;
+        }
+
+        QuicTraceLogConnVerbose(
+            DatagramSendQueued,
+            Connection,
+            "Datagram [%p] queued with %llu bytes (flags 0x%x)",
+            SendRequest,
+            SendRequest->TotalLength,
+            SendRequest->Flags);
+    }
+
+    if (Connection->State.PeerTransportParameterValid) {
+        QuicSendSetSendFlag(&Connection->Send, QUIC_CONN_SEND_FLAG_DATAGRAM);
+    }
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+BOOLEAN
+QuicDatagramWriteFrame(
+    _In_ QUIC_DATAGRAM* Datagram,
+    _Inout_ QUIC_PACKET_BUILDER* Builder
+    )
+{
+    QUIC_CONNECTION* Connection = QuicDatagramGetConnection(Datagram);
+    QUIC_DBG_ASSERT(Datagram->Enabled);
+
+    while (Datagram->SendQueue != NULL) {
+        QUIC_SEND_REQUEST* SendRequest = Datagram->SendQueue;
+        uint16_t AvailableBufferLength =
+            (uint16_t)Builder->Datagram->Length - Builder->EncryptionOverhead;
+
+        if (Builder->Metadata->Flags.KeyType == QUIC_PACKET_KEY_0_RTT &&
+            !(SendRequest->Flags & QUIC_SEND_FLAG_ALLOW_0_RTT)) {
+            //
+            // Not allowed to send the datagram in 0-RTT.
+            //
+            return FALSE;
+        }
+
+        BOOLEAN HadRoomForDatagram =
+            QuicDatagramFrameEncodeEx(
+                SendRequest->Buffers,
+                SendRequest->BufferCount,
+                SendRequest->TotalLength,
+                &Builder->DatagramLength,
+                AvailableBufferLength,
+                (uint8_t*)Builder->Datagram->Buffer);
+        if (!HadRoomForDatagram && Builder->Metadata->FrameCount != 0) {
+            //
+            // Part of the packet was used already and we didn't have room to frame
+            // this datagram. We need to try again in an unused packet.
+            //
+            return FALSE;
+        }
+
+        if (Datagram->PrioritySendQueueTail == &SendRequest->Next) {
+            Datagram->PrioritySendQueueTail = &Datagram->SendQueue;
+        }
+        if (Datagram->SendQueueTail == &SendRequest->Next) {
+            Datagram->SendQueueTail = &Datagram->SendQueue;
+        }
+        Datagram->SendQueue = SendRequest->Next;
+
+        if (HadRoomForDatagram) {
+            Builder->Metadata->Flags.IsRetransmittable = TRUE; // TODO - It's ack-eliciting but not really retransmittable

```

Comment:
Add new flag in follow up PR that makes that distinction?

Reply:
Really, we should just rename the flag to AckIliciting. But I don't want to do that right now.